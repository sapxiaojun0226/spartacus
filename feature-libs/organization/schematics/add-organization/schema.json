{
  "$schema": "http://json-schema.org/schema",
  "$id": "OrganizationSchematics",
  "title": "Organization Schematics",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "debug": {
      "description": "Display additional details during the running process.",
      "type": "boolean",
      "default": false
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the organization features.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "enum": ["Administration", "Order-Approval", "Account-Summary"],
        "type": "string"
      },
      "default": ["Administration", "Order-Approval", "Account-Summary"],
      "x-prompt": {
        "message": "Which Organization features would you like to set up?",
        "type": "list",
        "items": [
          {
            "value": "Administration",
            "label": "Administration (b2b feature)"
          },
          {
            "value": "Order-Approval",
            "label": "Order Approval (b2b feature)"
          },
          {
            "value": "Account-Summary",
            "label": "Account Summary (b2b feature, requires Organization - Administration)"
          }
        ]
      }
    }
  },
  "required": []
}
