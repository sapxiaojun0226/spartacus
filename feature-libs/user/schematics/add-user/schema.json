{
  "$schema": "http://json-schema.org/schema",
  "$id": "UserSchematics",
  "title": "User Schematics",
  "description": "Most of the Spartacus features require User Account feature to be properly configured.",
  "type": "object",
  "properties": {
    "project": {
      "type": "string",
      "description": "The name of the project.",
      "$default": {
        "$source": "projectName"
      }
    },
    "debug": {
      "description": "Display additional details during the running process.",
      "type": "boolean",
      "default": false
    },
    "lazy": {
      "type": "boolean",
      "description": "Lazy load the user features.",
      "default": true
    },
    "features": {
      "type": "array",
      "uniqueItems": true,
      "items": {
        "enum": ["User-Account", "User-Profile"],
        "type": "string"
      },
      "default": ["User-Account", "User-Profile"],
      "x-prompt": "Which features would you like to set up from the User library? Please note that for most Spartacus features to be properly configured, the User-Account feature is required."
    }
  },
  "required": []
}
