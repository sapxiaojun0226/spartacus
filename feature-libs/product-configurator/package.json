{
  "name": "@spartacus/product-configurator",
  "version": "4.1.0-next.0",
  "description": "Product configurator feature library for Spartacus",
  "keywords": [
    "spartacus",
    "framework",
    "storefront",
    "product configurator"
  ],
  "homepage": "https://github.com/SAP/spartacus",
  "repository": "https://github.com/SAP/spartacus/tree/develop/feature-libs/product-configurator",
  "license": "Apache-2.0",
  "exports": {
    ".": {
      "sass": "./_index.scss"
    }
  },
  "scripts": {
    "build:schematics": "yarn clean:schematics && ../../node_modules/.bin/tsc -p ./tsconfig.schematics.json",
    "clean:schematics": "../../node_modules/.bin/rimraf \"schematics/**/*.js\" \"schematics/**/*.js.map\" \"schematics/**/*.d.ts\"",
    "test:schematics": "yarn --cwd ../../projects/schematics/ run clean && yarn clean:schematics && ../../node_modules/.bin/jest --config ./jest.schematics.config.js"
  },
  "dependencies": {
    "tslib": "^2.4.0"
  },
  "peerDependencies": {
    "@angular-devkit/schematics": "^14.2.3",
    "@angular/common": "^14.2.3",
    "@angular/core": "^14.2.3",
    "@angular/forms": "^14.2.3",
    "@angular/router": "^14.2.3",
    "@ng-select/ng-select": "^9.0.2",
    "@ngrx/effects": "^14.3.0",
    "@ngrx/store": "^14.3.0",
    "@spartacus/cart": "4.1.0-next.0",
    "@spartacus/checkout": "4.1.0-next.0",
    "@spartacus/core": "4.1.0-next.0",
    "@spartacus/order": "4.2.0",
    "@spartacus/schematics": "4.1.0-next.0",
    "@spartacus/storefront": "4.1.0-next.0",
    "@spartacus/styles": "4.1.0-next.0",
    "rxjs": "^6.6.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "schematics": "./schematics/collection.json"
}
