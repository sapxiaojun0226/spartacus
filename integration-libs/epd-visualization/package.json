{
  "name": "@spartacus/epd-visualization",
  "version": "4.1.0-next.0",
  "description": "SAP Enterprise Product Development Visualization integration library for Spartacus",
  "keywords": [
    "spartacus",
    "framework",
    "storefront",
    "epd-visualization",
    "2d",
    "3d",
    "visual",
    "spare",
    "part",
    "picking",
    "epd"
  ],
  "homepage": "https://github.com/SAP/spartacus",
  "repository": "https://github.com/SAP/spartacus/tree/develop/integration-libs/epd-visualization",
  "license": "Apache-2.0",
  "exports": {
    ".": {
      "sass": "./_index.scss"
    }
  },
  "scripts": {
    "build:schematics": "yarn clean:schematics && ../../node_modules/.bin/tsc -p ./tsconfig.schematics.json",
    "clean:schematics": "../../node_modules/.bin/rimraf \"schematics/**/*.js\" \"schematics/**/*.js.map\" \"schematics/**/*.d.ts\"",
    "test:schematics": "yarn --cwd ../../projects/schematics/ run clean && yarn clean:schematics && ../../node_modules/.bin/jest --config ./jest.schematics.config.js"
  },
  "dependencies": {
    "tslib": "^2.4.0"
  },
  "peerDependencies": {
    "@angular-devkit/schematics": "^14.2.3",
    "@angular/common": "^14.2.3",
    "@angular/core": "^14.2.3",
    "@angular/forms": "^14.2.3",
    "@angular/router": "^14.2.3",
    "@sapui5/ts-types-esm": "1.98.0",
    "@spartacus/cart": "4.1.0-next.0",
    "@spartacus/core": "4.1.0-next.0",
    "@spartacus/schematics": "4.1.0-next.0",
    "@spartacus/storefront": "4.1.0-next.0",
    "@spartacus/styles": "4.1.0-next.0",
    "bootstrap": "^4.6.1",
    "rxjs": "^6.6.0"
  },
  "publishConfig": {
    "access": "public"
  },
  "schematics": "./schematics/collection.json"
}
