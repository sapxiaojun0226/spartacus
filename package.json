{
  "name": "storefrontapp",
  "version": "0.0.0",
  "license": "Apache-2.0",
  "author": "SAP, Spartacus team",
  "engines": {
    "node": "^14.15.0 || >=16.10.0"
  },
  "scripts": {
    "build:analyze": "yarn build --stats-json && npx --yes webpack-bundle-analyzer dist/storefrontapp/stats.json",
    "build": "env-cmd --no-override -e dev,b2c,$SPA_ENV ng build storefrontapp --configuration production",
    "build:core": "ng build core --configuration production",
    "build:storefrontlib": "ng build storefrontlib --configuration production",
    "build:asm": "yarn --cwd feature-libs/asm run build:schematics && ng build asm --configuration production",
    "build:user": "yarn --cwd feature-libs/user run build:schematics && ng build user --configuration production",
    "build:assets": "yarn --cwd ./projects/assets build",
    "build:cart": "yarn --cwd feature-libs/cart run build:schematics && ng build cart --configuration production",
    "build:cdc": "yarn --cwd integration-libs/cdc run build:schematics && ng build cdc --configuration production",
    "build:cds": "yarn --cwd integration-libs/cds run build:schematics && ng build cds --configuration production",
    "build:checkout": "yarn --cwd feature-libs/checkout run build:schematics && ng build checkout --configuration production",
    "build:digital-payments": "yarn --cwd integration-libs/digital-payments run build:schematics && ng build digital-payments --configuration production",
    "build:epd-visualization": "yarn --cwd integration-libs/epd-visualization run build:schematics && ng build epd-visualization --configuration production",
    "build:order": "yarn --cwd feature-libs/order run build:schematics && ng build order --configuration production",
    "build:libs": "ng build core --configuration production && ng build storefrontlib --configuration production && concurrently --kill-others-on-fail yarn:build:schematics yarn:build:user && yarn build:cart && concurrently --kill-others-on-fail yarn:build:order yarn:build:checkout yarn:build:asm yarn:build:tracking && concurrently --kill-others-on-fail yarn:build:organization yarn:build:product yarn:build:product-configurator yarn:build:storefinder yarn:build:cdc && concurrently --kill-others-on-fail yarn:build:smartedit yarn:build:qualtrics yarn:build:assets yarn:build:cds yarn:build:digital-payments yarn:build:epd-visualization && yarn build:setup",
    "build:organization": "yarn --cwd feature-libs/organization run build:schematics && ng build organization --configuration production",
    "build:product": "yarn --cwd feature-libs/product run build:schematics && ng build product --configuration production",
    "build:product-configurator": "yarn --cwd feature-libs/product-configurator run build:schematics && ng build product-configurator --configuration production",
    "build:qualtrics": "yarn --cwd feature-libs/qualtrics run build:schematics && ng build qualtrics --configuration production",
    "build:schematics": "yarn --cwd projects/schematics run build",
    "build:setup": "ng build setup --configuration production",
    "build:ssr": "env-cmd --no-override -e dev,b2c,$SPA_ENV ng run storefrontapp:server:production",
    "build:ssr:ci": "env-cmd -e ci,b2c,$SPA_ENV ng run storefrontapp:server:production",
    "build:storefinder": "yarn --cwd feature-libs/storefinder run build:schematics && ng build storefinder --configuration production",
    "build:smartedit": "yarn --cwd feature-libs/smartedit run build:schematics && ng build smartedit --configuration production",
    "build:tracking": "yarn --cwd feature-libs/tracking run build:schematics && ng build tracking --configuration production",
    "config:update": "ts-node ./tools/config/index.ts --fix",
    "config:check": "ts-node ./tools/config/index.ts",
    "generate:deps": "ts-node ./tools/schematics/dependency-collector.ts",
    "e2e:open": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:open",
    "e2e:open:b2b": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:open:b2b",
    "e2e:open:ccv2": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:open:ccv2",
    "e2e:run": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run",
    "e2e:run:b2b": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:b2b",
    "e2e:run:core": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:core",
    "e2e:run:ci": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci",
    "e2e:run:ci:ssr": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:ssr",
    "e2e:run:ci:b2b": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:b2b",
    "e2e:run:ci:flaky": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:flaky",
    "e2e:run:ci:core": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:core",
    "e2e:run:ci:core:b2b": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:core:b2b",
    "e2e:run:ci:cds": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:cds",
    "e2e:run:ci:product-configuration": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:product-configurator-vc",
    "e2e:run:ci:cdc": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:cdc",
    "e2e:run:ci:digital-payments": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:digital-payments",
    "e2e:run:ci:epd-visualization": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:epd-visualization",
    "e2e:run:ci:ccv2": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:ccv2",
    "e2e:run:ci:ccv2-b2b": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:ccv2-b2b",
    "e2e:run:ci:ccv2-cpq": "yarn --cwd ./projects/storefrontapp-e2e-cypress run cy:run:ci:ccv2-cpq",
    "generate:docs": "npx @compodoc/compodoc@1.1.13 -p tsconfig.compodoc.json && ./scripts/zip-docs.sh",
    "generate:publish:docs": "yarn generate:docs && yarn publish:docs",
    "i18n-lint": "i18n-lint -t \"{{,}}\" projects/storefrontlib/**/*.html feature-libs/**/*.html integration-libs/**/*.html -a alt,title,placeholder,aria-label",
    "lint": "ng lint",
    "lint:styles": "stylelint \"{projects,feature-libs}/**/*.scss\"",
    "prettier": "prettier --config ./.prettierrc --list-different \"{projects,feature-libs,core-libs,integration-libs}/**/*{.ts,.js,.json,.scss,.html}\"",
    "prettier:fix": "prettier --config ./.prettierrc --list-different \"{projects,feature-libs,core-libs,integration-libs}/**/*{.ts,.js,.json,.scss,.html}\" --write",
    "start": "env-cmd --no-override -e dev,b2c,$SPA_ENV ng serve",
    "start:b2b": "env-cmd --no-override -e dev,b2b,$SPA_ENV ng serve",
    "start:prod": "env-cmd --no-override -e dev,b2c,$SPA_ENV ng serve --configuration production",
    "start:pwa": "cd ./dist/storefrontapp/ && http-server --silent --proxy http://localhost:4200? -p 4200",
    "test": "ng test",
    "test:libs": "concurrently \"ng test core --code-coverage\" \"ng test storefrontlib --code-coverage\" \"ng test cart --code-coverage\" \"ng test organization --code-coverage\" \"ng test storefinder --code-coverage\" \"ng test smartedit --code-coverage\" \"ng test asm --code-coverage\" \"ng test qualtrics --code-coverage\" \"ng test product --code-coverage\" \"ng test product-configurator --code-coverage\" \"ng test cdc --code-coverage\" \"ng test setup --code-coverage\" \"ng test checkout --code-coverage\" \"ng test order --code-coverage\" \"ng test digital-payments --code-coverage\" \"ng test epd-visualization --code-coverage\"",
    "test:storefront:lib": "ng test storefrontlib --source-map --code-coverage",
    "dev:ssr": "env-cmd --no-override -e dev,b2c,$SPA_ENV cross-env NODE_TLS_REJECT_UNAUTHORIZED=0 ng run storefrontapp:serve-ssr",
    "serve:ssr": "node dist/storefrontapp-server/main.js",
    "serve:ssr:ci": "NODE_TLS_REJECT_UNAUTHORIZED=0 SSR_TIMEOUT=0 node dist/storefrontapp-server/main.js",
    "serve:ssr:dev": "cross-env NODE_TLS_REJECT_UNAUTHORIZED=0 node dist/storefrontapp-server/main.js",
    "prerender": "ng run storefrontapp:prerender --routesFile projects/storefrontapp/prerender.txt",
    "prerender:dev": "cross-env NODE_TLS_REJECT_UNAUTHORIZED=0 ng run storefrontapp:prerender --routesFile projects/storefrontapp/prerender.txt",
    "postinstall": "ngcc --properties es2015 browser module main --first-only --create-ivy-entry-points",
    "prerender:cds": "cross-env SPARTACUS_CDS=true ng run storefrontapp:prerender",
    "publish:docs": "gh-pages -d documentation -m \"Compodocs updates\"",
    "test:cds:lib": "ng test cds --code-coverage",
    "build:cds:ssr": "cross-env SPARTACUS_CDS=true ng run storefrontapp:server:production",
    "dev:cds:ssr": "cross-env NODE_TLS_REJECT_UNAUTHORIZED=0 SPARTACUS_CDS=true ng run storefrontapp:serve-ssr",
    "prerender:cds:dev": "cross-env NODE_TLS_REJECT_UNAUTHORIZED=0 SPARTACUS_CDS=true ng run storefrontapp:prerender --routesFile projects/storefrontapp/prerender.txt"
  },
  "private": false,
  "dependencies": {
    "@angular/animations": "^14.2.3",
    "@angular/common": "^14.2.3",
    "@angular/compiler": "^14.2.3",
    "@angular/core": "^14.2.3",
    "@angular/forms": "^14.2.3",
    "@angular/platform-browser": "^14.2.3",
    "@angular/platform-browser-dynamic": "^14.2.3",
    "@angular/platform-server": "^14.2.3",
    "@angular/pwa": "^14.2.3",
    "@angular/router": "^14.2.3",
    "@angular/service-worker": "^14.2.3",
    "@ng-select/ng-select": "^9.0.2",
    "@ngrx/effects": "^14.3.0",
    "@ngrx/router-store": "^14.3.0",
    "@ngrx/store": "^14.3.0",
    "@nguniversal/express-engine": "^14.0.0",
    "@types/google.maps": "^3.49.2",
    "angular-oauth2-oidc": "^13.0.1",
    "bootstrap": "^4.6.1",
    "comment-json": "^4.2.3",
    "express": "^4.15.2",
    "hamburgers": "^1.1.3",
    "i18next": "^21.9.1",
    "i18next-http-backend": "^1.4.1",
    "ngx-infinite-scroll": "^14.0.0",
    "parse5": "^6.0.1",
    "rxjs": "^6.6.0",
    "ts-loader": "^9.3.1",
    "tslib": "^2.4.0",
    "zone.js": "~0.11.8"
  },
  "devDependencies": {
    "@angular-builders/custom-webpack": "^14.0.1",
    "@angular-devkit/build-angular": "^14.2.3",
    "@angular-devkit/core": "^14.2.3",
    "@angular-devkit/schematics": "^14.2.3",
    "@angular-eslint/builder": "^14.1.2",
    "@angular-eslint/eslint-plugin": "^14.1.2",
    "@angular-eslint/eslint-plugin-template": "^14.1.2",
    "@angular-eslint/schematics": "^14.1.2",
    "@angular-eslint/template-parser": "^14.1.2",
    "@angular/cli": "^14.2.3",
    "@angular/compiler-cli": "^14.2.3",
    "@angular/language-service": "^14.2.3",
    "@babel/runtime": "^7.18.9",
    "@ngrx/store-devtools": "^14.3.0",
    "@nguniversal/builders": "^14.2.0",
    "@sapui5/ts-types-esm": "1.98.0",
    "@schematics/angular": "^14.2.3",
    "@spartacus-eslint/eslint-plugin": "link:./projects/eslint",
    "@types/express": "^4.17.0",
    "@types/fs-extra": "^9.0.7",
    "@types/i18next": "^12.1.0",
    "@types/jasmine": "~3.6.0",
    "@types/jasminewd2": "~2.0.3",
    "@types/jest": "^28.1.8",
    "@types/node": "^12.11.1",
    "@types/parse5": "^5.0.3",
    "@types/semver": "6",
    "@types/shelljs": "^0.8.7",
    "@typescript-eslint/eslint-plugin": "^5.36.1",
    "@typescript-eslint/parser": "^5.36.1",
    "chalk": "^4.1.0",
    "commander": "^3.0.0",
    "concurrently": "^5.3.0",
    "core-js": "^3.2.1",
    "cross-env": "^7.0.0",
    "enquirer": "^2.3.6",
    "env-cmd": "^10.1.0",
    "eslint": "^8.18.0",
    "eslint-plugin-deprecation": "^1.3.2",
    "eslint-plugin-import": "^2.26.0",
    "eslint-plugin-jsdoc": "^37.0.3",
    "eslint-plugin-prefer-arrow": "^1.2.3",
    "faker": "^4.1.0",
    "fs-extra": "^9.0.1",
    "gh-pages": "^2.2.0",
    "glob": "^7.1.6",
    "http-server": "^14.1.0",
    "i18n-lint": "^1.0.0",
    "jasmine-core": "~3.6.0",
    "jasmine-marbles": "^0.6.0",
    "jest": "^28.1.1",
    "jest-preset-angular": "^12.2.0",
    "jsonc-parser": "^3.0.0",
    "karma": "~6.3.16",
    "karma-chrome-launcher": "~3.1.0",
    "karma-cli": "^2.0.0",
    "karma-coverage": "~2.0.3",
    "karma-jasmine": "~4.0.0",
    "karma-jasmine-html-reporter": "^1.5.0",
    "karma-junit-reporter": "^2.0.1",
    "karma-parallel": "^0.3.1",
    "ng-packagr": "^14.2.0",
    "postcss": "^8.2.13",
    "postcss-scss": "^3.0.4",
    "prettier": "~2.4.1",
    "rimraf": "^3.0.2",
    "semver": "^7.3.5",
    "shelljs": "^0.8.5",
    "stylelint": "^13.13.1",
    "stylelint-scss": "^3.18.0",
    "ts-jest": "^28.0.8",
    "ts-morph": "^9.1.0",
    "ts-node": "^10.6.0",
    "typescript": "^4.7.2",
    "webpack": "^5.73.0",
    "webpack-cli": "^4.10.0"
  }
}
