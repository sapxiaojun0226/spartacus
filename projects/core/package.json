{
  "name": "@spartacus/core",
  "version": "4.1.0-next.0",
  "description": "Spartacus - the core framework",
  "keywords": [
    "spartacus",
    "storefront",
    "angular",
    "framework"
  ],
  "homepage": "https://github.com/SAP/spartacus",
  "repository": "https://github.com/SAP/spartacus/tree/develop/projects/core",
  "license": "Apache-2.0",
  "dependencies": {
    "tslib": "^2.4.0"
  },
  "peerDependencies": {
    "@angular/common": "^14.2.3",
    "@angular/core": "^14.2.3",
    "@angular/platform-browser": "^14.2.3",
    "@angular/router": "^14.2.3",
    "@ngrx/effects": "^14.3.0",
    "@ngrx/router-store": "^14.3.0",
    "@ngrx/store": "^14.3.0",
    "angular-oauth2-oidc": "^13.0.1",
    "i18next": "^21.9.1",
    "i18next-http-backend": "^1.4.1",
    "rxjs": "^6.6.0"
  },
  "publishConfig": {
    "access": "public"
  }
}
