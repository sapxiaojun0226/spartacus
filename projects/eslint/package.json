{
  "name": "@spartacus/eslint-plugin-eslint",
  "version": "0.0.0-dev",
  "private": true,
  "main": "index.js",
  "dependencies": {
    "tslib": "^2.4.0"
  }
}
