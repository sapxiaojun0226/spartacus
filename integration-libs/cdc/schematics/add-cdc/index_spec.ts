/// <reference types="jest" />

import {
  SchematicTestRunner,
  UnitTestTree,
} from '@angular-devkit/schematics/testing';
import {
  Schema as ApplicationOptions,
  Style,
} from '@schematics/angular/application/schema';
import { Schema as WorkspaceOptions } from '@schematics/angular/workspace/schema';
import {
  cdcFeatureModulePath,
  CDC_FEATURE_NAME,
  LibraryOptions as SpartacusCdcOptions,
  SpartacusOptions,
  SPARTACUS_ASM,
  SPARTACUS_CDC,
  SPARTACUS_SCHEMATICS,
  SPARTACUS_USER,
  userAccountWrapperModulePath,
  userProfileWrapperModulePath,
  USER_ACCOUNT_FEATURE_NAME,
  USER_PROFILE_FEATURE_NAME,
} from '@spartacus/schematics';
import * as path from 'path';
import { peerDependencies } from '../../package.json';

const collectionPath = path.join(__dirname, '../collection.json');

describe('Spartacus CDC schematics: ng-add', () => {
  const schematicRunner = new SchematicTestRunner(
    SPARTACUS_CDC,
    collectionPath
  );

  let appTree: UnitTestTree;

  const workspaceOptions: WorkspaceOptions = {
    name: 'workspace',
    version: '0.5.0',
  };

  const appOptions: ApplicationOptions = {
    name: 'schematics-test',
    inlineStyle: false,
    inlineTemplate: false,
    routing: false,
    style: Style.Scss,
    skipTests: false,
    projectRoot: '',
  };

  const spartacusDefaultOptions: SpartacusOptions = {
    project: 'schematics-test',
    lazy: true,
    features: [],
  };

  const libraryNoFeaturesOptions: SpartacusCdcOptions = {
    project: 'schematics-test',
    lazy: true,
    features: [],
  };

  const cdcFeatureOptions: SpartacusCdcOptions = {
    ...libraryNoFeaturesOptions,
    features: [CDC_FEATURE_NAME],
  };

  beforeEach(async () => {
    schematicRunner.registerCollection(
      SPARTACUS_SCHEMATICS,
      path.join(
        __dirname,
        '../../../../projects/schematics/src/collection.json'
      )
    );
    schematicRunner.registerCollection(
      SPARTACUS_ASM,
      require.resolve('../../../../feature-libs/asm/schematics/collection.json')
    );
    schematicRunner.registerCollection(
      SPARTACUS_USER,
      path.join(
        __dirname,
        '../../../../feature-libs/user/schematics/collection.json'
      )
    );

    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'workspace',
        workspaceOptions
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        '@schematics/angular',
        'application',
        appOptions,
        appTree
      )
      .toPromise();
    appTree = await schematicRunner
      .runExternalSchematicAsync(
        SPARTACUS_SCHEMATICS,
        'ng-add',
        { ...spartacusDefaultOptions, name: 'schematics-test' },
        appTree
      )
      .toPromise();
  });

  describe('Without features', () => {
    beforeEach(async () => {
      appTree = await schematicRunner
        .runSchematicAsync('ng-add', libraryNoFeaturesOptions, appTree)
        .toPromise();
    });

    it('should not create any of the feature modules', () => {
      expect(appTree.exists(cdcFeatureModulePath)).toBeFalsy();
    });
  });

  describe('CDC feature', () => {
    describe('validation of jsSDKUrl', () => {
      beforeEach(async () => {
        appTree = await schematicRunner
          .runSchematicAsync(
            'ng-add',
            {
              ...cdcFeatureOptions,
              features: [USER_ACCOUNT_FEATURE_NAME, USER_PROFILE_FEATURE_NAME],
            },
            appTree
          )
          .toPromise();
        appTree = await schematicRunner
          .runSchematicAsync(
            'ng-add',
            { ...cdcFeatureOptions, javascriptUrl: '<dc>.gigya.com/<api-key>' },
            appTree
          )
          .toPromise();
      });

      it('should set the given javascriptUrl', async () => {
        const featureModule = appTree.readContent(cdcFeatureModulePath);
        expect(featureModule).toMatchSnapshot();
      });
    });

    describe('general setup', () => {
      beforeEach(async () => {
        appTree = await schematicRunner
          .runSchematicAsync(
            'ng-add',
            {
              ...cdcFeatureOptions,
              features: [USER_ACCOUNT_FEATURE_NAME, USER_PROFILE_FEATURE_NAME],
            },
            appTree
          )
          .toPromise();
        appTree = await schematicRunner
          .runSchematicAsync('ng-add', cdcFeatureOptions, appTree)
          .toPromise();
      });

      it('should install necessary Spartacus libraries', () => {
        const packageJson = JSON.parse(appTree.readContent('package.json'));
        let dependencies: Record<string, string> = {};
        dependencies = { ...packageJson.dependencies };
        dependencies = { ...dependencies, ...packageJson.devDependencies };

        for (const toAdd in peerDependencies) {
          // skip the SPARTACUS_SCHEMATICS, as those are added only when running by the Angular CLI, and not in the testing environment
          if (
            !peerDependencies.hasOwnProperty(toAdd) ||
            toAdd === SPARTACUS_SCHEMATICS
          ) {
            continue;
          }
          // TODO: after 4.0: use this test, as we'll have synced versions between lib's and root package.json
          // const expectedVersion = (peerDependencies as Record<
          //   string,
          //   string
          // >)[toAdd];
          const expectedDependency = dependencies[toAdd];
          expect(expectedDependency).toBeTruthy();
          // expect(expectedDependency).toEqual(expectedVersion);
        }
      });

      it('should add the feature using the lazy loading syntax', async () => {
        const featureModule = appTree.readContent(cdcFeatureModulePath);
        expect(featureModule).toMatchSnapshot();
      });

      it('should install the appropriate dependencies', async () => {
        const userAccountWrapperModule = appTree.readContent(
          userAccountWrapperModulePath
        );
        expect(userAccountWrapperModule).toMatchSnapshot();

        const userProfileWrapperModule = appTree.readContent(
          userProfileWrapperModulePath
        );
        expect(userProfileWrapperModule).toMatchSnapshot();
      });
    });

    describe('eager loading', () => {
      beforeEach(async () => {
        appTree = await schematicRunner
          .runSchematicAsync(
            'ng-add',
            {
              ...cdcFeatureOptions,
              features: [USER_ACCOUNT_FEATURE_NAME, USER_PROFILE_FEATURE_NAME],
            },
            appTree
          )
          .toPromise();
        appTree = await schematicRunner
          .runSchematicAsync(
            'ng-add',
            { ...cdcFeatureOptions, lazy: false },
            appTree
          )
          .toPromise();
      });

      it('should import appropriate modules', async () => {
        const featureModule = appTree.readContent(cdcFeatureModulePath);
        expect(featureModule).toMatchSnapshot();
      });
    });
  });
});
