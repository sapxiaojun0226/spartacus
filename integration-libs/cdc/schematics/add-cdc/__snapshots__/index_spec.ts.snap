// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Spartacus CDC schematics: ng-add CDC feature eager loading should import appropriate modules 1`] = `
"import { NgModule } from '@angular/core';
import { CdcModule } from \\"@spartacus/cdc\\";
import { CdcConfig, CdcRootModule } from \\"@spartacus/cdc/root\\";
import { CDCUserAccountModule } from \\"@spartacus/cdc/user-account\\";
import { CDCUserProfileModule } from \\"@spartacus/cdc/user-profile\\";
import { provideConfig } from \\"@spartacus/core\\";

@NgModule({
  declarations: [],
  imports: [
    CdcRootModule,
    CdcModule,
    CDCUserAccountModule,
    CDCUserProfileModule
  ],
  providers: [provideConfig(<CdcConfig>{
    cdc: [
      {
        baseSite: 'electronics-spa',
        javascriptUrl: 'JS_SDK_URL_PLACEHOLDER',
        sessionExpiration: 3600
      },
    ],
  })]
})
export class CdcFeatureModule { }
"
`;

exports[`Spartacus CDC schematics: ng-add CDC feature general setup should add the feature using the lazy loading syntax 1`] = `
"import { NgModule } from '@angular/core';
import { CdcConfig, CdcRootModule, CDC_FEATURE } from \\"@spartacus/cdc/root\\";
import { CmsConfig, provideConfig } from \\"@spartacus/core\\";

@NgModule({
  declarations: [],
  imports: [
    CdcRootModule
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      [CDC_FEATURE]: {
        module: () =>
          import('@spartacus/cdc').then((m) => m.CdcModule),
      },
    }
  }),
  provideConfig(<CdcConfig>{
    cdc: [
      {
        baseSite: 'electronics-spa',
        javascriptUrl: 'JS_SDK_URL_PLACEHOLDER',
        sessionExpiration: 3600
      },
    ],
  })
  ]
})
export class CdcFeatureModule { }
"
`;

exports[`Spartacus CDC schematics: ng-add CDC feature general setup should install the appropriate dependencies 1`] = `
"import { NgModule } from '@angular/core';
import { CDCUserAccountModule } from \\"@spartacus/cdc/user-account\\";
import { UserAccountModule } from \\"@spartacus/user/account\\";

@NgModule({
  declarations: [],
  imports: [
    UserAccountModule,
    CDCUserAccountModule
  ]
})
export class UserAccountWrapperModule { }
"
`;

exports[`Spartacus CDC schematics: ng-add CDC feature general setup should install the appropriate dependencies 2`] = `
"import { NgModule } from '@angular/core';
import { CDCUserProfileModule } from \\"@spartacus/cdc/user-profile\\";
import { UserProfileModule } from \\"@spartacus/user/profile\\";

@NgModule({
  declarations: [],
  imports: [
    UserProfileModule,
    CDCUserProfileModule
  ]
})
export class UserProfileWrapperModule { }
"
`;

exports[`Spartacus CDC schematics: ng-add CDC feature validation of jsSDKUrl should set the given javascriptUrl 1`] = `
"import { NgModule } from '@angular/core';
import { CdcConfig, CdcRootModule, CDC_FEATURE } from \\"@spartacus/cdc/root\\";
import { CmsConfig, provideConfig } from \\"@spartacus/core\\";

@NgModule({
  declarations: [],
  imports: [
    CdcRootModule
  ],
  providers: [provideConfig(<CmsConfig>{
    featureModules: {
      [CDC_FEATURE]: {
        module: () =>
          import('@spartacus/cdc').then((m) => m.CdcModule),
      },
    }
  }),
  provideConfig(<CdcConfig>{
    cdc: [
      {
        baseSite: 'electronics-spa',
        javascriptUrl: '<dc>.gigya.com/<api-key>',
        sessionExpiration: 3600
      },
    ],
  })
  ]
})
export class CdcFeatureModule { }
"
`;
